reporters/kamon-opentelemetry/src/main/scala/kamon/otel/MetricsConverter.scala

```diff
@@ -0,0 +1,121 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kamon.otel
+
+import io.opentelemetry.sdk.common.InstrumentationScopeInfo
+import io.opentelemetry.sdk.metrics.data._
+import io.opentelemetry.sdk.metrics.internal.data._
+import io.opentelemetry.sdk.resources.Resource
+import kamon.metric.Instrument.Snapshot
+import kamon.metric.{Distribution, MeasurementUnit, MetricSnapshot, PeriodSnapshot}
+import kamon.tag.Lookups
+import kamon.trace.Span.TagKeys
+
+import java.lang.{Double => JDouble, Long => JLong}
+import java.time.Instant
+import java.util.{Collection => JCollection}
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+class WithResourceMetricsConverter(resource: Resource, kamonVersion: String, from: Instant, to: Instant) {
+  private val fromNs = from.toEpochMilli * 1000000
+  private val toNs = to.toEpochMilli * 1000000
+
+  private def instrumentationScopeInfo(snapshot: MetricSnapshot[_, _]): InstrumentationScopeInfo = {
+    // logic for looking up the component doesn't really seem to make sense - to be compliant we should probably be grouping the metrics by component before calling this
+    InstrumentationScopeInfo.create(snapshot.instruments.headOption.flatMap(_.tags.get(Lookups.option(TagKeys.Component))) getOrElse "kamon-instrumentation", kamonVersion, null)
+  }
+
+  private def toString(unit: MeasurementUnit): String = unit.magnitude.name
+
+  def toGaugeDatum(g: Snapshot[Double]): DoublePointData = ImmutableDoublePointData.create(fromNs, toNs, SpanConverter.toAttributes(g.tags), g.value)
+
+  def toGaugeData(g: Seq[Snapshot[Double]]): GaugeData[DoublePointData] = ImmutableGaugeData.create(g.map(toGaugeDatum).asJava)
+
+  def convertGauge(gauge: MetricSnapshot.Values[Double]): MetricData =
+    ImmutableMetricData.createDoubleGauge(
+      resource,
+      instrumentationScopeInfo(gauge),
+      gauge.name,
+      gauge.description,
+      toString(gauge.settings.unit),
+      toGaugeData(gauge.instruments))
+
+  def toHistogramDatum(s: Snapshot[Distribution]): HistogramPointData = {
+    val boundaries = ArrayBuffer.newBuilder[JDouble]
+    val counts = ArrayBuffer.newBuilder[JLong]
+    for (el <- s.value.bucketsIterator) {
+      counts += el.frequency
+      boundaries += el.value.toDouble
+    }
+    ImmutableHistogramPointData.create(
+      fromNs,
+      toNs,
+      SpanConverter.toAttributes(s.tags),
+      JDouble valueOf s.value.sum.toDouble,
+      JDouble valueOf s.value.min.toDouble,
+      JDouble valueOf s.value.max.toDouble,
+      boundaries.result().dropRight(1).asJava,
```

Comment:
Otel just wants bucket boundaries, so we drop the 'max' here

Reply:
Adding some comments I dropped on Discord for visibility:

Kamon's buckets are just HDR histogram buckets, which are pretty "small" buckets in comparison to what a usual Prometheus-like histogram expects. On Prom systems you would expect to expose 10-20 buckets per histogram, but Kamon's HDR can have thousands of buckets and that buckets iterator will only give you the buckets have values on the current tick, which might not be the same as in the previous tick

we do a bit of work on the prometheus exporter to re-bucket the data into a smaller resolution histogram based on the configured bucket boundaries and I think here we should do something similar BUT instead of doing that, I would rather go with the exponential histogram

I'm not sure if the exponential histograms are already part of the official spec and supported everywhere, but a bit of googling should clear that out